packages/rich-text/src/plugins/Paragraph/index.tsx

```diff
@@ -44,9 +55,29 @@ export function createParagraphPlugin(): PlatePlugin {
         element: element?.map((deserializeNode) => ({
           ...deserializeNode,
           deserialize: (el: HTMLElement) => {
-            if (el.nodeName === 'DIV' && el.hasAttribute('data-entity-type')) {
+            if (el.textContent === '' || isEmbed(el)) {
               return;
             }
+
+            // if any anchestor is a div that we convert into a paragraph or
+            // but not an embed, we exit to not have nested paragraphs
+            let currentParent = el.parentElement;
+            let hasEmbedAnchestor = false;
+            let hasParagraphAnchestor = false;
+            while (currentParent !== null) {
+              if (isEmbed(currentParent)) {
+                hasEmbedAnchestor = true;
+                break;
+              }
+              if (isParagraphDiv(currentParent)) {
+                hasParagraphAnchestor = true;
+              }
+              currentParent = currentParent.parentElement;
+            }
+            if (!hasEmbedAnchestor && hasParagraphAnchestor) {
+              return;
+            }
```

Comment:
Can't we solve this by using `editor.normalizeNodes`? I think we can use the rule of `['P', 'DIV']` and adding something like this to ensure only inline elements inside a paragraph:

```js
editor.normalizeNode = entry => {
    const [node, path] = entry
    // If the element is a paragraph, ensure its children are valid.
    if (Element.isElement(node) && node.type === BLOCKS.PARAGRAPH) {
      for (const [child, childPath] of Node.children(editor, path)) {
        if (Element.isElement(child) && !editor.isInline(child)) {
          Transforms.unwrapNodes(editor, { at: childPath })
          return
        }
      }
    }
    // Fall back to the original `normalizeNode` to enforce other constraints.
    normalizeNode(entry)
  }
```

Reply:
Sorry @cezarsmpio, I could get back at this only today. Yes normalization does the trick in a nicer way, so I used it.
I tried to play with the available options and method but non covered all the scenarios. Anyway the simple unwrap solves the bug while keeping the previous behaviors intact or improved.